### DISCUSSION TITLE
Questions on frameplacement

### DISCUSSION THREAD

**Discussion by xiaosuyang** [stack-of-tasks/pinocchio#2186] [Q&A] [answered]
Hello.
I found that both functions updateFramePlacement() and framesForwardKinematics() are used to calculate the placement of each frame. So what's the differences between them.

Besides,the model.frame[frame_id].placement return the placement with respect to parent joint, how to get the placement wrt the Base_link or world_frame?

Finally，for a legged robot, what's the position of the "world_frame“, is it a inertial frame whose origin coincide with base_frame(torso) or just a fixed point on the ground?

**Comment by stephane-caron** [comment, marked as answer]
>  I found that both functions updateFramePlacement() and framesForwardKinematics() are used to calculate the placement of each frame. So what's the differences between them.

From the documentation of [updateFramePlacement](https://gepettoweb.laas.fr/doc/stack-of-tasks/pinocchio/master/doxygen-html/namespacepinocchio.html#ab48760c9b4ad957cd95d6d3392b2c1e2) and [framesForwardKinematics](https://gepettoweb.laas.fr/doc/stack-of-tasks/pinocchio/master/doxygen-html/namespacepinocchio.html#a89903169c76d3c55bacaa2479bd39f76):

- For the former: "One of the algorithms forwardKinematics should have been called first" (``framesForwardKinematics`` is one of these algorithms). As a consequence there is no configuration vector `q` in the parameters, but there is a `frame_id` as it will only update the placement of one specific frame.
- For the latter: "First calls the forwardKinematics on the model, then computes the placement of each frame."

Therefore, we don't need to call `updateFramePlacement` if we call `framesForwardKinematics`, as it will update all frame placements (same effect as calling `updateFramePlacement` on every `frame_id`, although it will of course be done all at once and more efficiently).

This is only for your first question. Try opening a separate discussion for each question so that we can organize the conversation better.
